client/src/controllers/CloneController.ts

```diff
@@ -0,0 +1,148 @@
+  /** Internal tracking of whether a clearing animation is in progress. */
+  declare isClearingValue: boolean;
+  /** Delay, in milliseconds, before adjusting classes on show */
+  declare showDelayValue: number;
+
+  /**
+   * Adds a new element to the container based on the type argument provided in the event
+   * or action params objects. Optionally clearing the container first with support for
+   * added custom text inside the added element.
+   */
+  add(event?: CustomEvent<AddOptions> & { params?: AddOptions }) {
+    const {
+      clear = false,
+      text = '',
+      type = null,
+    } = { ...event?.detail, ...event?.params };
+
+    this.element.classList.add(...this.addedClasses);
+
+    if (clear) this.clear();
+
+    const content = this.getTemplateContent(type);
+    if (!content) return;
+
+    const textElement = content.lastElementChild;
+
+    if (textElement instanceof HTMLElement && text) {
+      textElement.textContent = text;
+    }
+
+    this.containerTarget.appendChild(content);
+
+    debounce(() => {
+      this.element.classList.remove(...this.hideClasses);
+      this.element.classList.add(...this.showClasses);
+      this.dispatch('added');
+    }, this.showDelayValue || null /* run immediately if zero */)();
+  }
+
+  /**
+   * If called with an event (or any truthy argument) reset the classes for show/hide
+   * so the this method can be used intentionally via actions allowing clearing after
+   * animations have run.
+   */
+  clear(event?: CustomEvent<ClearOptions> & { params?: ClearOptions }) {
+    this.isClearingValue = false;
+
+    if (!event) {
+      this.containerTarget.innerHTML = '';
+      return;
+    }
+
+    const element = this.element;
+
+    this.isClearingValue = true;
+
+    const { wait = 0 } = { ...event?.detail, ...event?.params };
+
+    const complete = () => {
+      if (!this?.isClearingValue) return;
+      this.containerTarget.innerHTML = '';
+      this.dispatch('cleared');
+      this.isClearingValue = false;
+    };
+
+    Promise.race([
+      wait &&
+        new Promise((resolve) => {
+          element.addEventListener('animationend', resolve, { once: true });
+        }),
+      debounce(noop, wait || null)(),
+    ]).then(complete);
+
+    element.classList.remove(...this.addedClasses);
+    element.classList.remove(...this.showClasses);
+    element.classList.add(...this.hideClasses);
+  }
+
+  /**
+   * If no type provided, return the first template target, otherwise try to find
+   * a matching target, finally fall back on the first template target if nothing
+   * is found.
+   */
+  getTemplateContent(type?: string | null): HTMLElement | null {
+    const template =
+      (type &&
+        this.templateTargets.find(({ dataset }) => dataset.type === type)) ||
+      this.templateTarget;
+    const content = template.content.firstElementChild?.cloneNode(true);
+    return content instanceof HTMLElement ? content : null;
+  }
+}
```

Comment:
Note: Unfortunately Git diffs do not do renamed files well. However the MessagesController has been renamed and while a lot has changed the core behaviour is still there. When reviewing the commits locally it should be a bit clearer what changed.